integrations/azure_ai_search/src/haystack_integrations/document_stores/azure_ai_search/document_store.py

```diff
@@ -0,0 +1,451 @@
+            Dictionary to deserialize from.
+
+        :returns:
+            Deserialized component.
+        """
+
+        deserialize_secrets_inplace(data["init_parameters"], keys=["api_key", "azure_endpoint"])
+        if (vector_search_configuration := data["init_parameters"].get("vector_search_configuration")) is not None:
+            data["init_parameters"]["vector_search_configuration"] = VectorSearch.from_dict(vector_search_configuration)
+        return default_from_dict(cls, data)
+
+    def count_documents(self) -> int:
+        """
+        Returns how many documents are present in the search index.
+
+        :returns: list of retrieved documents.
+        """
+        return self.client.get_document_count()
+
+    def write_documents(self, documents: List[Document], policy: DuplicatePolicy = DuplicatePolicy.FAIL) -> int:
+        """
+        Writes the provided documents to search index.
+
+        :param documents: documents to write to the index.
+        :return: the number of documents added to index.
+        """
+
+        def _convert_input_document(documents: Document):
+            document_dict = asdict(documents)
+            if not isinstance(document_dict["id"], str):
+                msg = f"Document id {document_dict['id']} is not a string, "
+                raise Exception(msg)
+            index_document = self._convert_haystack_documents_to_azure(document_dict)
+
+            return index_document
+
+        if len(documents) > 0:
+            if not isinstance(documents[0], Document):
+                msg = "param 'documents' must contain a list of objects of type Document"
+                raise ValueError(msg)
+
+        documents_to_write = []
+        for doc in documents:
+            try:
+                self.client.get_document(doc.id)
+                if policy == DuplicatePolicy.SKIP:
+                    logger.info(f"Document with ID {doc.id} already exists. Skipping.")
+                    continue
+                elif policy == DuplicatePolicy.FAIL:
+                    msg = f"Document with ID {doc.id} already exists."
+                    raise DuplicateDocumentError(msg)
+                elif policy == DuplicatePolicy.OVERWRITE:
+                    logger.info(f"Document with ID {doc.id} already exists. Overwriting.")
+                    documents_to_write.append(_convert_input_document(doc))
+            except ResourceNotFoundError:
+                # Document does not exist, safe to add
+                documents_to_write.append(_convert_input_document(doc))
+
+        if documents_to_write != []:
+            self.client.upload_documents(documents_to_write)
+        return len(documents_to_write)
+
+    def delete_documents(self, document_ids: List[str]) -> None:
+        """
+        Deletes all documents with a matching document_ids from the search index.
+
+        :param document_ids: ids of the documents to be deleted.
+        """
+        if self.count_documents == 0:
```

Comment:
We need to invoke count_documents method with parentheses here. This condition is never true.